### PR TITLE
Linux Networking Documentation

### PR DESCRIPTION
**Pull request by n-sandeep**
This PR includes:
-> moving DPDK specific documentation to dpdk directory -> Create ES2K specific linux networking documentation for IPv4.

### CODE REVIEW

**Review by ffoulkes — APPROVED** [review]
Spelling, grammar, and punctuation corrections noted.
I've also noted some markdown rule violations that a linter would catch.

**Review by ffoulkes — APPROVED** [review]
Approved on the grounds that the reorganization and the addition of a new document constitute an improvement the quality of the code base.

I'd appreciate it if you could address the spelling / grammar / punctuation errors before merging, or in an immediate follow-up, but I won't insist on it.

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+For each VM that is spawned for overlay network we need to have a port representer. We create VLAN netdev's on top of the IPDF netdev which is assigned to VSI group 3 in step-2 mentioned above.
+
+```
+ip link add link <VSI 8> name vlan1 type vlan id 1
+ip link add link <VSI 8> name vlan2 type vlan id 2
+ifconfig vlan1 up
+ifconfig vlan2 up
+```
+Note: Here the assumption is, we have created 2 overlay VM's and creating 2 port representers for those VM's.
+port representer should always bein format: `lower case string 'vlan'+'vlanID'`
```
Comment (nupuruttarwar) [inline p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:141]: Typo (bein) -> be in the format?

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+Note: Here we are creating VxLAN tunnel with VNI 0, user can create any VNI for tunneling.
+
+#### 7. Configure rules for overlay control packets.
+Configure rules to send overlay control pakcets from VM's to its respective port representers.
+
+Below configurations assumes
+- Overlay VF1 has a VSI value 14
+- Overlay VF2 has a VSI value 15
+
+These VSI values can be checked with `/usr/bin/cli_client -q -c` command on IMC. This command provides VSI ID, Vport ID and corresponding MAC addresses for all
```
Comment (nupuruttarwar) [inline p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:165]: Should we add a sample output from /usr/bin/cli_client -q -c?

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+- IDPF netdev's on ACC
+- VF's on HOST
+- IDPF netdev's on HOST (if IDPF driver loaded by user on HOST)
+- Netdev's on IMC
+
+```
+# Configuration for control packets from overlay VF with VSI-14 to VSI-8 by tagging a VLAN ID 1
```
Comment (nupuruttarwar) [inline p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:172]: Suggestion: Rules for control packets coming from overlay VF(VSI-14), IPU will add  a VLAN tag 1 and send to HOST1(VSI-8)?

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+ip route change 40.1.1.0/24 via 40.1.1.2 dev <IDPF netdev for VSI 10>
+```
+
+#### 11. Test the ping scenarios:
+  - Ping between VM's on the same host
+  - Underlay ping
+  - Overlay ping: Ping between VM's on different hosts
+
+## Limitations <a name="limitations"></a>
+Current SAI enablement for the networking recipe has following limitations:
```
Comment (nupuruttarwar) [inline p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:233]: these limitations are for linux networking P4 and not networking recipe?

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+- All VLAN interfaces created on top of IDPF netdev, should always be in lowercase format "vlan+vlan_id"
+Ex: vlan1, vlan2, vlan3 ... vlan4094
+- br-int port, vxlan0 port and adding vlan ports to br-int need to be done after loading the pipeline.
```
Comment (nupuruttarwar) [inline p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:236]: Easier way to say this would be to "Set the pipeline before adding ...?"

**Review by nupuruttarwar — APPROVED** [review]
Minor comments and suggestions. Please fix all mdl errors before merging

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+#### 1. Set the forwarding pipeline:
+Once the application is started, set the forwarding pipeline config using
+P4Runtime Client `p4rt-ctl` set-pipe command
+
+```bash
+$P4CP_INSTALL/bin/p4rt-ctl set-pipe br0 $OUTPUT_DIR/linux_networking.pb.bin $OUTPUT_DIR/linux_networking.p4info.txt
+```
+Note: Assuming `linux_networking.pb.bin` and `linux_networking.p4info.txt` along with other P4 artifacts are created as per the steps mentioned in previous section.
+
+#### 2. Configure VSI Group and add a netdev :
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:63]: Sure, This is taken care from Derek comments, yet to upload a patch

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+# SEM_DIRECT_MAP_PGEN_CTRL: LSB 11-bit is for vsi which need to map into vsig
+devmem 0x20292002a0 64 0xA000050000000008
+```
+
+Note: Here VSI 8 has been used for receiving all control packets and added to VSI group 3.
+Modify this VSI based on your configuration.
+
+#### 3. Create Overlay network:
+Create VF's on HOST and spawn VM's on top of those VF's.
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:81]: I didnt get this comment, can we discuss offline?

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md (hunk 4 above)
Comment (n-sandeep) [reply p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:233]: Done, rephrased it

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md (hunk 2 above)
Comment (n-sandeep) [reply p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:165]: That will lead to few other queries on other params from that CLI o/p.
If user is trying out LNW, they should be familiar to basic bringup and how to re-use the VSI info.

File: p4src/linux_networking/es2k/README_LINUX_NETWORKING.md [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+Example to create 4 VF's:  echo 4 > /sys/devices/pci0000:ae/0000:ae:00.0/0000:af:00.0/sriov_numvfs
+```
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply p4src/linux_networking/es2k/README_LINUX_NETWORKING.md:83]: somehow i cant view this comment..

**Comment by n-sandeep** [comment]
will raise a seperate PR for ECMP and IPV6

**Pull request merged**